README.md
[![codecov](https://codecov.io/gh/center-for-threat-informed-defense/attack-control-framework-mappings/branch/master/graph/badge.svg?token=PLVBGOUWMC)](https://codecov.io/gh/center-for-threat-informed-defense/attack-control-framework-mappings)

# Security Control Framework Mappings to ATT&CK
This repository contains security control framework mappings to MITRE ATT&CK® with supporting documentation and resources. These mappings provide a critically important resource for organizations to assess their security control coverage against real-world threats as described in the ATT&CK knowledge base and provide a foundation for integrating ATT&CK-based threat information into the risk management process. This work was developed by the [Center for Threat-Informed Defense](https://mitre-engenuity.org/center-for-threat-informed-defense/) in collaboration with our participants.

**NIST 800-53 Revision 4 Security Control Mappings**

| ATT&CK Version | Mappings as XLSX (download) | ATT&CK Navigator Layers | STIX Data |
|---|---|---|---|
| [ATT&CK-v9.0](/frameworks/ATT&CK-v9.0/nist800-53-r4/) | [Spreadsheet](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/raw/develop/frameworks/ATT%26CK-v9.0/nist800-53-r4/nist800-53-r4-mappings.xlsx) | [Navigator Layers](/frameworks/ATT&CK-v9.0/nist800-53-r4/layers) | [STIX](/frameworks/ATT&CK-v9.0/nist800-53-r4/stix) |
| [ATT&CK-v8.2](/frameworks/ATT&CK-v8.2/nist800-53-r4/) | [Spreadsheet](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/raw/develop/frameworks/ATT%26CK-v8.2/nist800-53-r4/nist800-53-r4-mappings.xlsx) | [Navigator Layers](/frameworks/ATT&CK-v8.2/nist800-53-r4/layers) | [STIX](/frameworks/ATT&CK-v8.2/nist800-53-r4/stix) |

**NIST 800-53 Revision 5 Security Control Mappings**

| ATT&CK Version | Mappings as XLSX (download) | ATT&CK Navigator Layers | STIX Data |
|---|---|---|---|
| [ATT&CK-v9.0](/frameworks/ATT&CK-v9.0/nist800-53-r5/) | [Spreadsheet](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/raw/develop/frameworks/ATT%26CK-v9.0/nist800-53-r5/nist800-53-r5-mappings.xlsx) | [Navigator Layers](/frameworks/ATT&CK-v9.0/nist800-53-r5/layers) | [STIX](/frameworks/ATT&CK-v9.0/nist800-53-r5/stix) |
| [ATT&CK-v8.2](/frameworks/ATT&CK-v8.2/nist800-53-r5/) | [Spreadsheet](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/raw/develop/frameworks/ATT%26CK-v8.2/nist800-53-r5/nist800-53-r5-mappings.xlsx) | [Navigator Layers](/frameworks/ATT&CK-v8.2/nist800-53-r5/layers) | [STIX](/frameworks/ATT&CK-v8.2/nist800-53-r5/stix) |

## A Collaborative Approach

Mapping NIST Special Publication 800-53, or any security control framework, to ATT&CK is a labor intensive and often subjective undertaking. Furthermore, due to the large number of security controls in any given framework and the evolving nature of cyber adversaries, these mappings are often error prone and difficult to maintain. We recognized that there was not only a need for mappings NIST 800-53, but an opportunity to work collaboratively and advance threat-informed defense with the global community. With over 6,300 individual mappings between NIST 800-53 and ATT&CK, we believe that this work will greatly reduce the burden on the community – allowing organizations to focus their limited time and resources on understanding how controls map to threats in their environment.

## Repository Contents

- [Frameworks](/frameworks) — this directory contains the security control frameworks and their mappings to ATT&CK techniques. Each security control framework has its own directory of documentation and resources. 
- [Mapping Methodology](/docs/mapping_methodology.md) — a description of the general process used to create the control mappings
- [Tooling](/docs/tooling.md) — a set of python tools to support the creation of new mappings and the customization of existing mappings
- [Use Cases](/docs/use-cases.md) - use cases for security control framework mappings to ATT&CK
- [STIX Format](/docs/STIX_format.md) — information regarding the STIX representation of the control frameworks and the mappings to ATT&CK
- [Visualization](/docs/visualization.md) — this document describes some ways the mappings data can be visualized. 
- [Contributing](/CONTRIBUTING.md) — information about how to contribute controls, mappings, or other improvements to this repository
- [Changelog](/CHANGELOG.md) — list of updates to this repository


## Getting Involved

There are several ways that you can get involved with this project and help advance threat-informed defense. 

First, review the mappings, use them, and tell us what you think. We welcome your review and feedback on the NIST 800-53 mappings, our methodology, and resources. 

Second, we are interested in applying our methodology to other security control frameworks. Let us know what frameworks you would like to see mapped to ATT&CK. Your input will help us prioritize how we expand our mappings. 

Finally, we are interested developing additional tools and resources to help the community understand and make threat-informed decisions in their risk management programs. Share your ideas and we will consider them as we explore additional research projects.  

## Questions and Feedback
   
Please submit issues for any technical questions/concerns or contact ctid@mitre-engenuity.org directly for more general inquiries.

Also see the guidance for contributors if are you interested in [contributing or simply reporting issues.](/CONTRIBUTING.md)

## Notice 

Copyright 2020 MITRE Engenuity. Approved for public release. Document number CT0011

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at 

http://www.apache.org/licenses/LICENSE-2.0 

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License. 

This project makes use of ATT&CK®

[ATT&CK Terms of Use](https://attack.mitre.org/resources/terms-of-use/)


CHANGELOG.md
<!--    CHANGELOG FORMAT                                                -->

<!--    Completed Entry template:                                       -->
<!--                                                                    -->
<!--    # Date in DD MMM YYYY format                                    -->
<!--    ### New Features                                                -->
<!--    ### Improvements                                                -->
<!--    ### Fixes                                                       -->

<!--    Entries for pull request template:                              -->
<!--                                                                    -->
<!--    # Changes staged on develop                                     -->
<!--    ### New Features                                                -->
<!--    ### Improvements                                                -->
<!--    ### Fixes                                                       -->

# 19 July 2021
## nist800-53-r4 v2.0 and nist800-53-r5 v2.0
### Fixes
- Release of the security control mappings and tools updated with ATT&CK-v9.0, including layers and stix data
- Update project structure to hold prior mappings created for this repo (ATT&CK-v8.2)
- Update project structure and scripts, code quality updates, tests, coverage


# 18 May 2021
### Fixes
- NIST 800-53 R4 layers updated with attack-navigator 4.3. See issue [#65](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/issues/65).
- NIST 800-53 R5 layers updated with attack-navigator 4.3. See issue [#65](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/issues/65).
- mappingsToHeatmaps.py also updated to indicate attack-navigator version 4.3.

# 5 April 2021
## nist800-53-r5 v1.3
### Fixes
- Fixed a bug where subcontrol-of relationships were not being created between controls and enhancements in nist-800-53-r5. See issue [#61](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/issues/61).
- Updates to mappings in nist-800-53-r5 to address withdrawn controls (SA family) and to remove policy control (XX-1) mappings.

# 3 February 2021
## nist800-53-r4 v1.2 and nist800-53-r5 v1.2
### Fixes
- Fixes parse_mappings.py for nist800-53-r4 and nist800-53-r5 to remove duplicate mappings. See issue [#58](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/issues/58).

# 7 January 2021
## nist800-53-r4 v1.1 and nist800-53-r5 v1.1
### Fixes
- Fixed broken regex which was leading to erroneous mappings, in particular mappings to control enhancements. See issue [#56](https://github.com/center-for-threat-informed-defense/attack-control-framework-mappings/issues/56).

# 15 December 2020
## nist800-53-r4 v1.0 and nist800-53-r5 v1.0
- Initial release of security control framework mapping methodology and tools. 
- Initial release of NIST 800-53 R4 mappings to ATT&CK version 8.1.
- Initial release of NIST 800-53 R5 mappings to ATT&CK version 8.1.


src/list_mappings.py
import argparse
import json
import os

from colorama import Fore
from stix2.v20 import Bundle
import openpyxl
import openpyxl.utils
import pandas
import requests


def mappings_to_df(attack_bundle, controls_bundle, mappings_bundle):
    """Return a pandas dataframe listing the mappings in mappings_bundle"""
    rows = []
    for mapping in mappings_bundle.objects:
        control = controls_bundle.get(mapping.source_ref)
        if not control:
            print(Fore.RED + f"ERROR: cannot find object with ID {mapping.source_ref} in controls bundle" + Fore.RESET)
            exit()
        else:
            control = control[0]

        technique = attack_bundle.get(mapping.target_ref)
        if not technique:
            print(Fore.RED + f"ERROR: cannot find object with ID {mapping.target_ref} in ATT&CK bundle" + Fore.RESET)
            exit()
        else:
            technique = technique[0]

        rows.append({
            "Control ID": control["external_references"][0]["external_id"],
            "Control Name": control["name"],
            "Mapping Type": mapping["relationship_type"],
            "Technique ID": technique["external_references"][0]["external_id"],
            "Technique Name": technique["name"],
        })

    data_frame = pandas.DataFrame(rows)
    data_frame.sort_values(['Control ID', 'Technique ID'], ascending=[True, True], inplace=True)

    return data_frame


def workbook_changes(filename):
    """Changes spreadsheet format width, freezes first row, and sets
    filtering reference"""
    sheet_name = 'Sheet1'
    freeze_row = 'A2'  # freezes the first row of the document

    control_id_width = 14
    control_name_width = 69
    mapping_type_width = 18
    technique_id_width = 18
    technique_name_width = 58

    column_widths = [
        control_id_width,
        control_name_width,
        mapping_type_width,
        technique_id_width,
        technique_name_width,
    ]

    workbook = openpyxl.load_workbook(filename)
    worksheet = workbook[sheet_name]
    worksheet.freeze_panes = worksheet[freeze_row]

    # establishes filtering references in document
    auto_filter_section = f'A1:E{len(list(worksheet.rows))}'
    worksheet.auto_filter.ref = auto_filter_section

    for i, column_width in enumerate(column_widths):
        worksheet.column_dimensions[openpyxl.utils.get_column_letter(i + 1)].width = column_width

    workbook.save(filename)


if __name__ == "__main__":
    extensionToPDExport = {
        "xlsx": "to_excel",  # extension to df export function name
        "csv": "to_csv",
        "html": "to_html",
        "md": "to_markdown",
    }
    allowedExtensionList = ", ".join(extensionToPDExport.keys())
    parser = argparse.ArgumentParser(description="List mappings in human readable formats")
    parser.add_argument("-controls",
                        dest="controls",
                        help="filepath to the stix bundle representing the control framework",
                        default=os.path.join("..", "frameworks", "ATT&CK-v9.0", "nist800-53-r4",
                                             "stix", "nist800-53-r4-controls.json"))
    parser.add_argument("-mappings",
                        dest="mappings",
                        help="filepath to the stix bundle mapping the controls to ATT&CK",
                        default=os.path.join("..", "frameworks", "ATT&CK-v9.0", "nist800-53-r4",
                                             "stix", "nist800-53-r4-mappings.json"))
    parser.add_argument("-domain",
                        dest="domain",
                        help="which ATT&CK domain to use",
                        default="enterprise-attack")
    parser.add_argument("-version",
                        dest="version",
                        help="which ATT&CK version to use",
                        default="v9.0")
    parser.add_argument("-output",
                        help=f"filepath to write the output mappings to. Output format will be "
                             f"inferred from the extension. Allowed extensions: {allowedExtensionList}",
                        default=os.path.join("..", "frameworks", "ATT&CK-v9.0", "nist800-53-r4",
                                             "nist800-53-r4-mappings.xlsx"))

    args = parser.parse_args()

    if args.version != "v9.0":
        args.controls = args.controls.replace("ATT&CK-v9.0", f"ATT&CK-{args.version}")
        args.mappings = args.mappings.replace("ATT&CK-v9.0", f"ATT&CK-{args.version}")
        args.output = args.output.replace("ATT&CK-v9.0", f"ATT&CK-{args.version}")

    extension = args.output.split(".")[-1]
    if extension not in extensionToPDExport:
        msg = (f"ERROR: Unknown output extension \"{extension}\", please make "
               f"sure your output extension is one of: {allowedExtensionList}")
        print(Fore.RED + msg, Fore.reset)
        exit()

    print("downloading ATT&CK data... ", end="", flush=True)
    url = f"https://raw.githubusercontent.com/mitre/cti/ATT%26CK-{args.version}/{args.domain}/{args.domain}.json"
    attack_data = Bundle(
        requests.get(url, verify=True).json()["objects"],
        allow_custom=True
    )
    print("done")

    print("loading controls framework... ", end="", flush=True)
    with open(args.controls, "r") as f:
        controls = Bundle(json.load(f)["objects"], allow_custom=True)
    print("done")

    print("loading mappings... ", end="", flush=True)
    with open(args.mappings, "r") as f:
        mappings = Bundle(json.load(f)["objects"])
    df = mappings_to_df(attack_data, controls, mappings)
    print("done")

    print(f"writing {args.output}...", end="", flush=True)
    if extension in ["md"]:  # md doesn't support index=False and requires a stream and not a path
        with open(args.output, "w") as f:
            getattr(df, extensionToPDExport[extension])(f)
    else:
        getattr(df, extensionToPDExport[extension])(args.output, index=False)

        if extension in ["xlsx"]:
            workbook_changes(args.output)

        print("done")
